#27 implement database manager and connect

kortov opened the pull request:
connect to #27

kortov committed:
#27:  added DatabaseManager class

kortov committed:
#27:  added DatabaseManagerTests and improved DatabaseManager

ivshapovalov reviewed (approved):
Well done. . Test method names is good.

AVM82 commented on src/main/java/juja/sqlcmd/DatabaseManager.java:
```diff
@@ -0,0 +1,38 @@
+        this.jdbcUrl = jdbcUrl;
+    }
+
+    boolean connect(String database, String user, String password) {
```
Please make this method public.

kortov replied:
Done

kortov committed:
#27:  minor changes. Made connect method public

kortov committed:
#27:  minor changes. Reorganised import statements

Shtramak commented on src/main/java/juja/sqlcmd/DatabaseManager.java:
```diff
@@ -5,16 +5,33 @@
+        jdbcUrl = "jdbc:postgresql://localhost:5432/";
+    }
+
+    public DatabaseManager(String jdbcDriverClass, String jdbcUrl) {
```
Who will use this constructor to create instance of DatabaseManager()? Should client know the details about jdbcDriverClass or jdbcUrl? In this case I think it's better to have jdbcDriverClass and jdbcUrl declared as a constants.

kortov replied:
It's just for testing of bad url and driverName)

Shtramak commented on src/main/java/juja/sqlcmd/DatabaseManager.java:
```diff
@@ -5,16 +5,33 @@
 
 public class DatabaseManager {
+    private final String jdbcDriverClass;
+    private final String jdbcUrl;
```
Pay your attention on whitespaces (http://bit.ly/2j2qPE1)

Shtramak commented on src/main/java/juja/sqlcmd/DatabaseManager.java:
```diff
@@ -5,16 +5,33 @@
-            Class.forName("org.postgresql.Driver");
+            Class.forName(jdbcDriverClass);
+            if (connection != null) {
+                connection.close();
```
Maybe it's better to notify the client somehow that already there is unclosed connection? That's the moment when you provide additional logic to your implementation without requirements in a task. (http://bit.ly/2jRFHEn)

kortov replied:
I just copied it from Baglai's example but actually despite these lines look reasonable, I don't see any hanging connections in db without them. May be it's redundant

Shtramak commented on src/main/java/juja/sqlcmd/DatabaseManager.java:
```diff
@@ -5,16 +5,33 @@
             return true;
         } catch (ClassNotFoundException | SQLException e) {
             e.printStackTrace();
+            connection = null;
```
What if connection is not closed at this moment?

kortov replied:
Good question, I just copied it from Baglai's example)

Shtramak commented on src/test/java/juja/sqlcmd/DatabaseManagerTest.java:
```diff
@@ -0,0 +1,66 @@
+        originalErr = System.err;
+        System.setOut(new PrintStream(outContent));
+        System.setErr(new PrintStream(errContent));
+        databaseManager = new DatabaseManager();
```
databaseManager = new DatabaseManager(); is the only line you need in this method. Why you change streams from System? And then it's better to rename this method to init() or setup()

kortov replied:
There will be many stacktraces in tests of DbManager constructor, I just hide them

kortov replied:
About init method :) https://youtu.be/otrfSgeK3JI?t=601

Shtramak commented on src/test/java/juja/sqlcmd/DatabaseManagerTest.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @After
+    public void cleanUpStreams() {
```
You don't need this method after deleting unnesessary lines from setUpStreamsAndDbManager()

kortov committed:
#17: minor changes acc. to reviews

- changed access to dbManager ctor to package
- removed redundant lines from connect method
